Client/Frontend/Brave Today/Composer/FeedDataSource+RSS.swift

```diff
@@ -0,0 +1,230 @@
+// Copyright 2020 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import Data
+import FeedKit
+import Fuzi
+import Shared
+
+struct RSSFeedLocation: Hashable {
+    var title: String?
+    var url: URL
+    
+    var id: String {
+        url.absoluteString
+    }
+}
+
+extension FeedDataSource {
+    
+    // MARK: - RSS Sources
+    
+    var rssFeedLocations: [RSSFeedLocation] {
+        RSSFeedSource.all().compactMap {
+            guard let url = URL(string: $0.feedUrl) else { return nil }
+            return RSSFeedLocation(title: $0.title, url: url)
+        }
+    }
+    
+    /// Add a users custom RSS feed to the list of sources
+    ///
+    /// - returns: `true` if the feed is successfully added, `false` if it already exists or the
+    ///            url location is not a web page url
+    @discardableResult
+    func addRSSFeedLocation(_ location: RSSFeedLocation) -> Bool {
+        if !location.url.isWebPage(includeDataURIs: false) {
+            return false
+        }
+        let feedUrl = location.url.absoluteString
+        if RSSFeedSource.get(with: feedUrl) != nil {
+            return false
+        }
+        RSSFeedSource.insert(title: location.title,
+                             feedUrl: feedUrl)
+        setNeedsReloadCards()
+        return true
+    }
+    
+    /// Remove a users custom RSS feed to the list of sources
+    func removeRSSFeed(with url: URL) {
+        let feedUrl = url.absoluteString
+        if RSSFeedSource.get(with: feedUrl) == nil {
+            return
+        }
+        RSSFeedSource.delete(with: url.absoluteString)
+        setNeedsReloadCards()
+    }
+    
+    /// Whether or not an RSS feed is currently enabled
+    ///
+    /// - note: RSS Feeds are enabled by default since they are added by the user
+    func isRSSFeedEnabled(_ location: RSSFeedLocation) -> Bool {
+        FeedSourceOverride.get(fromId: location.id)?.enabled ?? true
+    }
+    
+    /// Toggle an RSS feed enabled state
+    func toggleRSSFeedEnabled(_ location: RSSFeedLocation, enabled: Bool) {
+        FeedSourceOverride.setEnabled(forId: location.id, enabled: enabled)
+        setNeedsReloadCards()
+    }
+}
+
+extension FeedItem.Content {
+    private static func imageURL(from document: HTMLDocument, releativeTo baseURL: URL?) -> URL? {
+        if let src = document.firstChild(xpath: "//img[@src]")?.attr("src"),
+           let url = URL(string: src, relativeTo: baseURL),
```

Comment:
nit: convert it to a util function?

Reply:
Convert what to a util function? This is the util function.